pkg/engine2/path_selection/candidate_validity.go

```diff
@@ -246,7 +246,7 @@ func (d downstreamChecker) makeValid(resource, operationResource *construct.Reso
 		var errs error
 		rt, err := d.ctx.KnowledgeBase().GetResourceTemplate(r)
 		if err != nil || rt == nil {
-			return false, fmt.Errorf("error getting resource template for resource %s: %w", resource, err)
+			return false, fmt.Errorf("error getting resource template for resource %s: %w", resource.ID, err)
```

Comment:
side comment - we should add a `String()` to resource that just prints its ID. This is a common enough occurrence and if printing properties is desired that can be implemented via `%+v` or something.

Reply:
yeah not a bad idea